pkg/api/types.go

```diff
@@ -1783,11 +1789,15 @@ type Toleration struct {
 	// +optional
 	Value string
 	// Effect indicates the taint effect to match. Empty means match all taint effects.
-	// When specified, allowed values are NoSchedule and PreferNoSchedule.
+	// When specified, allowed values are NoSchedule, PreferNoSchedule and NoExecute.
 	// +optional
 	Effect TaintEffect
-	// TODO: For forgiveness (#1574), we'd eventually add at least a grace period
-	// here, and possibly an occurrence threshold and period.
+	// TolerationSeconds represents the period of time the toleration (which must be
+	// of effect NoExecute, otherwise this field is ignored) tolerates the taint. By default,
+	// it is not set, which means tolerate the taint forever (do not evict). Zero and
+	// negative values are not allowed.
+	// +optional
+	TolerationSeconds *int64
```

Comment:
I remember we were discussing it already, but why can't this be a plain value (instead of ptr), with the convention that negative values mean infinity? For backward compatibility we only care that 0 value has a good default semantics (I'm not sure if it's "don't tolerate at all" or "tolerate forever")